jerry-core/lit/lit-strings.c

```diff
@@ -31,89 +40,82 @@ lit_is_valid_utf8_string (const lit_utf8_byte_t *utf8_buf_p, /**< utf-8 string *
                           lit_utf8_size_t buf_size, /**< string size */
                           bool is_strict) /**< true if surrogate pairs are not allowed */
 {
-  lit_utf8_size_t idx = 0;
+  const unsigned char *end = buf_size + utf8_buf_p;
 
-  bool is_prev_code_point_high_surrogate = false;
-  while (idx < buf_size)
+  const unsigned char *idx = (const unsigned char *) utf8_buf_p;
+
+  while (idx < end)
   {
-    lit_utf8_byte_t c = utf8_buf_p[idx++];
-    if ((c & LIT_UTF8_1_BYTE_MASK) == LIT_UTF8_1_BYTE_MARKER)
+    const uint8_t first_byte = *idx++;
+
+    if (first_byte < LIT_UTF8_EXTRA_BYTE_MARKER)
     {
-      is_prev_code_point_high_surrogate = false;
       continue;
     }
 
-    lit_code_point_t code_point = 0;
-    lit_code_point_t min_code_point = 0;
-    lit_utf8_size_t extra_bytes_count;
-    if ((c & LIT_UTF8_2_BYTE_MASK) == LIT_UTF8_2_BYTE_MARKER)
+    if (first_byte < LIT_UTF8_OVERLONG_MARKER || idx >= end)
     {
-      extra_bytes_count = 1;
-      min_code_point = LIT_UTF8_2_BYTE_CODE_POINT_MIN;
-      code_point = ((uint32_t) (c & LIT_UTF8_LAST_5_BITS_MASK));
-    }
-    else if ((c & LIT_UTF8_3_BYTE_MASK) == LIT_UTF8_3_BYTE_MARKER)
-    {
-      extra_bytes_count = 2;
-      min_code_point = LIT_UTF8_3_BYTE_CODE_POINT_MIN;
-      code_point = ((uint32_t) (c & LIT_UTF8_LAST_4_BITS_MASK));
+      return false;
     }
-    else if ((c & LIT_UTF8_4_BYTE_MASK) == LIT_UTF8_4_BYTE_MARKER)
+
+    const uint8_t second_byte = *idx++;
+
+    if ((second_byte & LIT_UTF8_EXTRA_BYTE_MASK) != LIT_UTF8_EXTRA_BYTE_MARKER)
     {
-      extra_bytes_count = 3;
-      min_code_point = LIT_UTF8_4_BYTE_CODE_POINT_MIN;
-      code_point = ((uint32_t) (c & LIT_UTF8_LAST_3_BITS_MASK));
+      return false;
     }
-    else
+
+    if (first_byte < LIT_UTF8_3_BYTE_MARKER)
     {
-      /* utf-8 string could not contain 5- and 6-byte sequences. */
-      return false;
+      continue;
     }
 
-    if (idx + extra_bytes_count > buf_size)
+    if (idx >= end || ((*idx++ & LIT_UTF8_EXTRA_BYTE_MASK) != LIT_UTF8_EXTRA_BYTE_MARKER))
     {
-      /* utf-8 string breaks in the middle */
       return false;
     }
 
-    for (lit_utf8_size_t offset = 0; offset < extra_bytes_count; ++offset)
+    if (first_byte < LIT_UTF8_4_BYTE_MARKER)
     {
-      c = utf8_buf_p[idx + offset];
-      if ((c & LIT_UTF8_EXTRA_BYTE_MASK) != LIT_UTF8_EXTRA_BYTE_MARKER)
+      if ((first_byte == LIT_UTF8_3_BYTE_MARKER && (second_byte & LIT_UTF8_2_BYTE_MASK) == LIT_UTF8_EXTRA_BYTE_MARKER))
       {
-        /* invalid continuation byte */
         return false;
       }
-      code_point <<= LIT_UTF8_BITS_IN_EXTRA_BYTES;
-      code_point |= (c & LIT_UTF8_LAST_6_BITS_MASK);
+
+      if (first_byte == LIT_UTF8_SURROGATE_MARKER
+          && second_byte >= LIT_UTF8_HIGH_SURROGATE_MIN
+          && second_byte <= LIT_UTF8_HIGH_SURROGATE_MAX
+          && *idx++ == LIT_UTF8_SURROGATE_MARKER)
+      {
+        const uint8_t fifth_byte = *idx++;
```

Comment:
fifth_byte? I am not sure everything is checked here.

This should looks like this:

First check the third byte, its value is irrelevant.

`if (is_strict && first_byte == surrogate_marker && second_byte in surrogate && idx + 3 <= end && check the other 3 bytes) -> return false;`